pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancerAdminToolClusterIntegrationTest.java

```diff
@@ -0,0 +1,868 @@
+        final int port = CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT + i;
+        final String instanceID = serverPrefix + port;
+        idealStateCopy.setPartitionState(segment, instanceID, "ONLINE");
+      }
+    }
+
+    if (llc) {
+      // write 2 latest segments for the same kafka partitions (0 and 1)
+      // with higher sequence number and in CONSUMING state
+      final String timestamp = idealStateMap.keySet().iterator().next().split(SegmentName.SEPARATOR)[3];
+      final String segmentPrefix = tableName + "__";
+      final String segmentSuffix = "__"+ timestamp;
+      int sequenceNumber = 1;
+      for (int kafkaPartition = 0; kafkaPartition < 2; kafkaPartition++) {
+        final String segmentID = segmentPrefix + kafkaPartition + "__" + sequenceNumber + segmentSuffix;
+        for (int j = 0; j < numReplicas; j++) {
+          final int port = CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT + j;
+          final String instanceID = serverPrefix + port;
+          idealStateCopy.setPartitionState(segmentID, instanceID, "CONSUMING");
+        }
+      }
+    }
+
+    final ZkBaseDataAccessor zkBaseDataAccessor = (ZkBaseDataAccessor) dataAccessor.getBaseDataAccessor();
+    zkBaseDataAccessor.set(idealStateKey.getPath(), idealStateCopy.getRecord(), idealState.getRecord().getVersion(),
+        AccessOption.PERSISTENT);
+  }
+
+  private void createIdealState(final int numSegments, final int numReplicas,
+      final String tableName)
+      throws Exception {
+    final HelixDataAccessor dataAccessor = _helixManagers.get(0).getHelixDataAccessor();
+    final String tableNameWithType = tableName + "_OFFLINE";
+    final PropertyKey idealStateKey = dataAccessor.keyBuilder().idealStates(tableNameWithType);
+    final IdealState idealState = dataAccessor.getProperty(idealStateKey);
+    final IdealState idealStateCopy = HelixHelper.cloneIdealState(idealState);
+    final String serverPrefix = CommonConstants.Helix.PREFIX_OF_SERVER_INSTANCE + NetUtil.getHostAddress() + "_";
+    for (int i = 0; i < numSegments; i++) {
+      final String segmentID = "segment" + i;
+      if (idealStateCopy.getInstanceStateMap(segmentID) != null) {
+        idealStateCopy.getInstanceStateMap(segmentID).clear();
+      }
+      for (int j = 0; j < numReplicas; j++) {
+        final int port = CommonConstants.Helix.DEFAULT_SERVER_NETTY_PORT + j;
+        final String instanceID = serverPrefix + port;
+        idealStateCopy.setPartitionState(segmentID, instanceID, "ONLINE");
+      }
+    }
+    // this CAS serves no purpose in this single threaded execution test
+    final ZkBaseDataAccessor zkBaseDataAccessor = (ZkBaseDataAccessor) dataAccessor.getBaseDataAccessor();
+    zkBaseDataAccessor.set(idealStateKey.getPath(), idealStateCopy.getRecord(), idealState.getRecord().getVersion(),
+        AccessOption.PERSISTENT);
+  }
+
+  public class FakeServerSegmentStateModelFactory extends StateModelFactory<StateModel> {
```

Comment:
would be a good thing to re-use methods to add/stop fake instances in ControllerTest (jackie cleaned up the tests recently), but we can leave that to a later PR. ControllerTest is a super-class of TableRebalancerAdminToolClusterIntegrationTest
Can you please add a TODO here?

Reply:
Done